library/packages/src/lib/y2packager/zypp_url.rb

```diff
@@ -0,0 +1,157 @@
+# ------------------------------------------------------------------------------
+# Copyright (c) 2017-2020 SUSE LLC, All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of version 2 of the GNU General Public License as published by the
+# Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+# ------------------------------------------------------------------------------
+
+require "uri"
+require "yast"
+require "delegate"
+
+module Y2Packager
+  # This class represents a libzypp URL
+  #
+  # Libzypp URLs do not conform to rfc3986 because they can include the so-called
+  # Repository Variables. Those vars can have several formats like $variable,
+  # ${variable}, ${variable-word} or ${variable+word}. They can appear in any
+  # component of the URL (host, path, port...) except in scheme, user or password.
+  #
+  # See https://doc.opensuse.org/projects/libzypp/HEAD/zypp-repovars.html
+  #
+  # The current implementation relies on SimpleDelegator to expose all the methods
+  # of an underlying URI object, so objects of this class can be used as a direct
+  # replacement in places that used to use URI, like {Y2Packager::Repository#raw_url}
+  #
+  class ZyppUrl < SimpleDelegator
+    Yast.import "Pkg"
+
+    # Repository schemes considered local (see #local?)
+    # https://github.com/openSUSE/libzypp/blob/a7a038aeda1ad6d9e441e7d3755612aa83320dce/zypp/Url.cc#L458
+    LOCAL_SCHEMES = [:cd, :dvd, :dir, :hd, :iso, :file].freeze
+
+    # Constructor
+    #
+    # @param url [String, ZyppUrl, URI::Generic]
+    def initialize(url)
+      __setobj__(URI(repovars_escape(url.to_s)))
+    end
+
+    # @return [URI::Generic]
+    def uri
+      __getobj__
+    end
+
+    alias_method :to_uri, :uri
+
+    # Constructs String from the URL
+    #
+    # @return [String]
+    def to_s
+      repovars_unescape(uri.to_s)
+    end
+
+    # See URI::Generic#hostname
+    #
+    # @return [String]
+    def hostname
+      repovars_unescape(uri.hostname)
+    end
+
+    # See URI::Generic#hostname
+    #
+    # TODO: escaping does not work here because the port wouldn't accept the
+    # escaped characters either. We likely need to modify the regexp used by
```

Comment:
Although is true that they say

> Flexibility to have an alternate URI::Parser (or just different patterns and regexp's).  — https://ruby-doc.org/stdlib-2.7.1/libdoc/uri/rdoc/URI.html

It isn't clear to me that we can provide to the parser the regexp to check the port. If I'm not wrong, only `ESCAPED`, `UNRESERVED`, `RESERVED`, `DOMLABEL`, `TOPLABEL`, `HOSTNAME` regexps [can be _redefined_](https://github.com/ruby/uri/blob/41c79e77628b2c65acc51e6a652641db490eaf8a/lib/uri/rfc2396_parser.rb#L334-L339), but not the one used to [check the `PORT`](https://github.com/ruby/uri/blob/41c79e77628b2c65acc51e6a652641db490eaf8a/lib/uri/rfc2396_parser.rb#L394).